File: 132_LongestWordWithPrefixes.py
"""
Problem: Longest String with All Prefixes
Description: Find the longest string where all prefixes exist as complete words.

Approach: Build trie with all words. DFS to find longest word where all prefixes
          are complete words (marked as end).
"""

from typing import List

class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()
    
    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isEnd = True

def dfs(node: TrieNode, current: str, result: List[str]):
    if len(current) > len(result[0]) or \
       (len(current) == len(result[0]) and current < result[0]):
        result[0] = current
    
    for c in sorted(node.children.keys()):
        if node.children[c].isEnd:
            dfs(node.children[c], current + c, result)

def longestWord(words: List[str]) -> str:
    trie = Trie()
    for word in words:
        trie.insert(word)
    
    result = [""]
    dfs(trie.root, "", result)
    
    return result[0]

def main():
    words1 = ["a", "banana", "app", "appl", "ap", "apply", "apple"]
    print("Longest word:", longestWord(words1))  # Output: apple
    
    words2 = ["abc", "bc", "ab", "qwe"]
    print("Longest word:", longestWord(words2))

if __name__ == "__main__":
    main()

"""
Time Complexity: O(n * m) where n is number of words, m is max length
Space Complexity: O(n * m)
"""

File: 132_LongestWordWithPrefixes.cpp
/*
Problem: Longest String with All Prefixes
Description: Find the longest string where all prefixes exist as complete words.

Approach: Build trie with all words. DFS to find longest word where all prefixes
          are complete words (marked as end).
*/

#include <bits/stdc++.h>
using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    bool isEnd;
    
    TrieNode() {
        for (int i = 0; i < 26; i++)
            children[i] = nullptr;
        isEnd = false;
    }
};

class Trie {
public:
    TrieNode* root;
    
    Trie() {
        root = new TrieNode();
    }
    
    void insert(string word) {
        TrieNode* node = root;
        for (char c : word) {
            int idx = c - 'a';
            if (!node->children[idx])
                node->children[idx] = new TrieNode();
            node = node->children[idx];
        }
        node->isEnd = true;
    }
};

void dfs(TrieNode* node, string current, string& result) {
    if (current.size() > result.size() || 
        (current.size() == result.size() && current < result)) {
        result = current;
    }
    
    for (int i = 0; i < 26; i++) {
        if (node->children[i] && node->children[i]->isEnd) {
            current.push_back('a' + i);
            dfs(node->children[i], current, result);
            current.pop_back();
        }
    }
}

string longestWord(vector<string>& words) {
    Trie trie;
    for (const string& word : words) {
        trie.insert(word);
    }
    
    string result = "";
    dfs(trie.root, "", result);
    
    return result;
}

int main() {
    vector<string> words1 = {"a", "banana", "app", "appl", "ap", "apply", "apple"};
    cout << "Longest word: " << longestWord(words1) << endl;  // Output: apple
    
    vector<string> words2 = {"abc", "bc", "ab", "qwe"};
    cout << "Longest word: " << longestWord(words2) << endl;
    
    return 0;
}

/*
Time Complexity: O(n * m) where n is number of words, m is max length
Space Complexity: O(n * m)
*/
